### PR TITLE
Feature/minimal variant price sorting

### PR DESCRIPTION
**Pull request by derenio**
This `PR`:
-  adds `minimal_variant_price` to the `Product` model,
    * a field that stores the minimal price among all the variants of the product,
- updates the `minimal_variant_price` on:
    * product's creation/update,
    * product variant's creation/update/removal,
    * sale's creation/update.

### CODE REVIEW

**Comment by NyanKiyoshi** [comment]
Did you notice that pytest is printing 100 time:
> .tox/py36-django22/lib/python3.6/site-packages/django/db/models/fields/__init__.py:1363

I believe it's a warning, but I don't see any message. Could you take a look?

File: saleor/discount/models.py [hunk 1]
```diff
@@ -149,12 +149,16 @@ def validate_min_amount_spent(self, value):
 
 
 class SaleQueryset(models.QuerySet):
-    def active(self, date):
+    def active(self, date=None):
+        if date is None:
+            date = timezone.now().date()
```
Comment (fowczarek) [inline saleor/discount/models.py]: Start and End date in Sale model are DateTimeField we should use 
`timezone.now()` instead of `timezone.now().date()`

File: saleor/discount/models.py [hunk 2]
```diff
@@ -149,12 +149,16 @@ def validate_min_amount_spent(self, value):
...
         return self.filter(
             Q(end_date__isnull=True) | Q(end_date__gte=date), start_date__lte=date
         )
 
-    def expired(self, date):
+    def expired(self, date=None):
+        if date is None:
+            date = timezone.now().date()
```
Comment (fowczarek) [inline saleor/discount/models.py]: Start and End date in Sale model are DateTimeField we should use 
`timezone.now()` instead of `timezone.now().date()`

File: saleor/discount/utils.py [hunk 3]
```diff
@@ -137,3 +137,7 @@ def fetch_discounts(date: datetime.date):
         )
         for sale in sales
     ]
+
+
+def fetch_active_discounts():
+    return fetch_discounts(datetime.date.today())
```
Comment (fowczarek) [inline saleor/discount/utils.py]: We should use `timezone.now()` instead of `datetime.date.today()`

File: saleor/product/migrations/0096_product_minimal_variant_price.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Generated by Django 2.2.1 on 2019-06-10 11:23
+
+import django_prices.models
+from django.db import migrations
+
+
+def populate_product_minimal_variant_price(apps, schema_editor):
+    Product = apps.get_model("product", "Product")
+    for product in Product.objects.prefetch_related("variants").iterator():
```
Comment (korycins) [inline saleor/product/migrations/0096_product_minimal_variant_price.py]: First of all. AFAIK `prefetch_related` and `iterator` will not work together. 
Secondly - Do you need to fetch `variants` here?

File: saleor/product/tasks.py [hunk 5]
```diff
@@ -29,3 +34,14 @@ def update_variants_names(product_type_pk, saved_attributes_ids):
...
     return _update_variants_names(instance, saved_attributes)
+
+
+@app.task
+def update_all_products_minimal_variant_prices_task():
+    update_products_minimal_variant_prices()
+
+
+@app.task
+def update_products_minimal_variant_prices_of_discount_task(discount_pk):
```
Comment (korycins) [inline saleor/product/tasks.py]: Are you gonna run these function somewhere?

File: saleor/product/utils/variant_prices.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+from ...discount.utils import fetch_active_discounts
+from ..models import Product
+
+
+def _get_product_minimal_variant_price(product, discounts):
+    # Start with the product's price as the minimal one
+    minimal_variant_price = product.price
+    for variant in product.variants.all():
+        variant_price = variant.get_price(discounts=discounts)
```
Comment (korycins) [inline saleor/product/utils/variant_prices.py]: `get_price` will return normal `Money`. No need to cast it. All `taxes` logic has been extracted from `get_price`

File: saleor/product/utils/variant_prices.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    minimal_variant_price = _get_product_minimal_variant_price(product, discounts)
+    if product.minimal_variant_price != minimal_variant_price:
+        product.minimal_variant_price = minimal_variant_price
+        product.save(update_fields=["minimal_variant_price"])
+    return product
+
+
+def update_products_minimal_variant_prices(products=None, discounts=None):
+    if products is None:
+        products = Product.objects.prefetch_related("variants").iterator()
```
Comment (korycins) [inline saleor/product/utils/variant_prices.py]: it won't work.
**Only prefetch**
```
Product.objects.prefetch_related('variants')[0].variants.all()
DEBUG django.db.backends (0.001) SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC  LIMIT 1; args=() [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC  LIMIT 1; args=() [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" IN (72); args=(72,) [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" IN (72); args=(72,) [PID:4554:MainThread]
Out[5]: <ProductVariantQueryset [<ProductVariant: 2l>, <ProductVariant: 500ml>, <ProductVariant: 1l>]>
next(Product.objects.prefetch_related('variants').iterator()).variants.all()
```

**Prefetch +iterator**
```
next(Product.objects.prefetch_related('variants').iterator()).variants.all()
DEBUG django.db.backends (0.001) DECLARE "_django_curs_4687144384_1" NO SCROLL CURSOR WITH HOLD FOR SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC; args=() [PID:4554:MainThread]
DEBUG django.db.backends (0.001) DECLARE "_django_curs_4687144384_1" NO SCROLL CURSOR WITH HOLD FOR SELECT "product_product"."id", "product_product"."publication_date", "product_product"."is_published", "product_product"."seo_title", "product_product"."seo_description", "product_product"."product_type_id", "product_product"."name", "product_product"."description", "product_product"."description_json", "product_product"."category_id", "product_product"."price", "product_product"."minimal_variant_price", "product_product"."attributes", "product_product"."updated_at", "product_product"."charge_taxes", "product_product"."weight", "product_product"."meta" FROM "product_product" ORDER BY "product_product"."name" ASC; args=() [PID:4554:MainThread]
Out[6]: DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:MainThread]
DEBUG django.db.backends (0.001) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:MainThread]
<ProductVariantQueryset [<ProductVariant: 2l>, <ProductVariant: 500ml>, <ProductVariant: 1l>]>
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-59]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-59]
DEBUG django.db.backends (0.005) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-56]
DEBUG django.db.backends (0.005) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-56]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-62]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-62]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-66]
DEBUG django.db.backends (0.002) SELECT "product_productvariant"."id", "product_productvariant"."sku", "product_productvariant"."name", "product_productvariant"."price_override", "product_productvariant"."product_id", "product_productvariant"."attributes", "product_productvariant"."track_inventory", "product_productvariant"."quantity", "product_productvariant"."quantity_allocated", "product_productvariant"."cost_price", "product_productvariant"."weight" FROM "product_productvariant" WHERE "product_productvariant"."product_id" = 72  LIMIT 21; args=(72,) [PID:4554:Thread-66]
```

File: tests/api/test_product.py [hunk 8]
```diff
@@ -740,7 +743,7 @@ def test_create_product(
     product_is_published = True
     product_charge_taxes = True
     product_tax_rate = "STANDARD"
-    product_price = 22.33
+    product_price = "22.33"
```
Comment (korycins) [inline tests/api/test_product.py]: Why did you change it to `str`? It should work with float

File: tests/test_product_minimal_variant_price.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    data = {
+        "name": "Product name",
+        "description": "Description.",
+        "price": "9.99",
+        "category": category.pk,
+    }
+
+    response = admin_client.post(url, data)
+    assert response.status_code == 302
+    [product] = Product.objects.all()  # Also checks there is only one product
```
Comment (korycins) [inline tests/test_product_minimal_variant_price.py]: simpler - `assert Product.objects.count() == 1`

**Comment by django-queries** [comment]
Here is the report for f2d0ec20e7a682e7d0288f9bc430ef3319f51601 (mirumee/saleor @ feature/minimal_variant_price_sorting)
Base comparison is ee9b5ac51231de5f6c78beb0c39756a4e08645c3.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  add billing address to checkout     	         34	         34	             20
  add shipping to checkout            	          7	          7	              0
  checkout payment charge             	         14	         14	              0
  complete checkout                   	          6	          6	              0
  create checkout                     	         43	         43	             21

# api.benchmark homepage
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve main menu                  	          5	          5	              0
  retrieve product list               	          4	          4	              0
  retrieve secondary menu             	          5	          5	              0
  retrieve shop                       	          2	          2	              0

# api.benchmark product
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  product details                     	         13	         13	              3

# api.benchmark variant
  test name                           	left count 	right count	duplicate count
  ------------------------------------	-----------	-----------	---------------
  retrieve variant list               	          9	          9	              2
```

</p>
</details>

File: saleor/product/tasks.py (hunk 5 above)
Comment (maarcingebala) [reply saleor/product/tasks.py]: This is used when creating/editing discounts in Dashboard 1.0. @derenio What about support for updating the minimal variant price in API? That's should be our top priority.

File: saleor/product/tasks.py [hunk 10]
```diff
@@ -29,3 +35,27 @@ def update_variants_names(product_type_pk, saved_attributes_ids):
...
+
+
+@app.task
+def update_products_minimal_variant_prices_task(product_pks):
+    products = Product.objects.filter(pk__in=product_pks)
+    update_products_minimal_variant_prices(products)
+
+
+@app.task
+def update_all_products_minimal_variant_prices_task():
```
Comment (maarcingebala) [inline saleor/product/tasks.py]: Is this task used?

Comment (derenio) [reply saleor/product/tasks.py]: No. I've created this one with the intend to plug it in the future to e.g. celery-beat with a periodic task.

With the introduction of a `Datetime` as the `Sale's` _start_ and _end_ dates we would need something more granular, ideally something that would update the `minimal variant prices` at the exact time the sale's start/end.

File: saleor/product/utils/variant_prices.py (hunk 7 above)
Comment (derenio) [reply saleor/product/utils/variant_prices.py]: I've removed the `prefetch_related` code.

File: tests/api/test_product.py (hunk 8 above)
Comment (derenio) [reply tests/api/test_product.py]: @korycins these are money amounts and the `Product.price` is a `MoneyField` (`Decimal` beneath) that can accept string representations of decimal numbers. Floats aren't accurate and work here only because the amount is properly rounded while you cast it to string (the default precision is less than 17 decimal places):
```python
[ins] In [1]: f"{22.33:.17}"                                                                        
Out[1]: '22.329999999999998'
```

It's a good habit to pass strings to `Decimals` and generally use them instead of floats when one deals with money.

File: saleor/product/tasks.py (hunk 5 above)
Comment (derenio) [reply saleor/product/tasks.py]: @maarcingebala I've added code to update the `minimal_variant_price` while creating/updating/deleting `Products`, `ProductVariants`, `Collections` and `Sales`.
We don't have to worry about `Categories` as these cascade delete products on removal. Also only products can be added to categories and we already handle that.

**Comment by koradon** [comment]
@derenio please rebase this PR to master

**Comment by koradon** [comment]
@derenio please fix CI

File: saleor/graphql/discount/mutations.py [hunk 11]
```diff
@@ -33,6 +37,14 @@ class BaseDiscountCatalogueMutation(BaseMutation):
     class Meta:
         abstract = True
 
+    @classmethod
+    def recalculate_discounts(cls, products, categories, collections):
```
Comment (korycins) [inline saleor/graphql/discount/mutations.py]: It recalculates prices, not exactly discounts. Am I right? `discounts` associates with `Discount` model

File: saleor/product/utils/variant_prices.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        product.minimal_variant_price = minimal_variant_price
+        product.save(update_fields=["minimal_variant_price"])
+    return product
+
+
+def update_products_minimal_variant_prices(products, discounts=None):
+    if discounts is None:
+        discounts = fetch_active_discounts()
+    for product in products:
+        update_product_minimal_variant_price(product, discounts)
```
Comment (korycins) [inline saleor/product/utils/variant_prices.py]: Instead of making `.save()` after `.save()` we could do single `bulk_update` at the end of function. https://docs.djangoproject.com/en/2.2/ref/models/querysets/#bulk-update
It will be chaper.

File: saleor/product/migrations/0102_merge_20190802_1110.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+# Generated by Django 2.2.3 on 2019-08-02 16:10
```
Comment (korycins) [inline saleor/product/migrations/0102_merge_20190802_1110.py]: Do we need that? It is not released yet so I assume that you can remove your migration and create the new one.

File: saleor/product/tasks.py [hunk 14]
```diff
@@ -31,3 +38,30 @@ def update_variants_names(product_type_pk, saved_attributes_ids):
...
+
+
+@app.task
+def update_products_minimal_variant_prices_of_discount_task(discount_pk):
+    discount = Sale.objects.get(pk=discount_pk)
+    update_products_minimal_variant_prices_of_discount(discount)
+
+
+@app.task
+def update_all_products_minimal_variant_prices_task():
```
Comment (korycins) [inline saleor/product/tasks.py]: Do we need this task? All products will always have the up to date `minimal_variant_prices`. So there is no need to have this task. Am I right?

File: tests/api/test_product_minimal_variant_price.py [hunk 15]
```diff
@@ -0,0 +1,606 @@
...
+    mock_update_minimal_variant_prices_task.delay.assert_called_once_with(
+        product_ids=[p.pk for p in product_list]
+    )
+
+
+@patch(
+    "saleor.graphql.product.mutations"
+    ".products.update_products_minimal_variant_prices_of_catalogues_task"
+)
+def test_CollectionRemoveProducts_updates_minimal_variant_price(
```
Comment (korycins) [inline tests/api/test_product_minimal_variant_price.py]: I don't think that we have convention with camel case for API classes.
`test_collection_remove_products_updates_minimal_variant_price`

**Comment by maarcingebala** [comment]
@derenio Please rebase this branch with `master` as we've merged a large PR yesterday. Then I'll be able to re-review it.

File: saleor/graphql/discount/mutations.py (hunk 11 above)
Comment (derenio) [reply saleor/graphql/discount/mutations.py]: The recalculated `minimal price` includes discounts hence the name.
For clarity I've renamed it to `recalculate_minimal_price`.

File: saleor/product/utils/variant_prices.py (hunk 12 above)
Comment (derenio) [reply saleor/product/utils/variant_prices.py]: Good idea! Changed.

File: saleor/product/migrations/0102_merge_20190802_1110.py (hunk 13 above)
Comment (derenio) [reply saleor/product/migrations/0102_merge_20190802_1110.py]: While rebaseing I've removed the merge migration and but my on top of the latest one.

File: saleor/product/tasks.py (hunk 14 above)
Comment (derenio) [reply saleor/product/tasks.py]: I thought so myself but then I've recalled that sales can expire. Even worse they can now expire or start and any hour of the day!

We need to establish default way for saleor to spawn periodical tasks, then we can track what Sales' start/end in the refresh window and recalculate affected product's minimal prices. However, I would still run this `update all` task once ~at midnight just to be sure.

File: tests/api/test_product_minimal_variant_price.py (hunk 15 above)
Comment (derenio) [reply tests/api/test_product_minimal_variant_price.py]: Test names should tell you what it's about and in these ones I'm testing particular GraphQL endpoints that class names I'm using here. This way the test is more readable. You can straight away know what part of the test's name is the tested part and which one the expected behavior.

**Comment by derenio** [comment]
@maarcingebala just rebased the `PR`.

File: saleor/graphql/product/schema.py [hunk 16]
```diff
@@ -201,6 +201,18 @@ class ProductQueries(graphene.ObjectType):
         price_gte=graphene.Float(
             description="Filter by price greater than or equal to the given value."
         ),
+        minimal_price_lte=graphene.Float(
```
Comment (maarcingebala) [inline saleor/graphql/product/schema.py]: We should now only use `filter` object to keep all available filters. These two fields that we add here are not necessary. `price_gte` and `price_lte` should also be removed but I'm not sure if it isn't used currently by our storefront.

**Comment by maarcingebala** [comment]
How do we handle published vs. unpublished collections in terms of minimal prices? E.g., if there is a sale for a particular collection, but the collection isn't published yet. Can we even create a sale for unpublished collections? I'm not sure how it works now but I'm writing that down for us not to forget to check it.

**Comment by koradon** [comment]
Publish/unpublish and calculating of minimal price are for me two separate things.
We should be able to publish/unpublish every variant we want and calculate the minimal price if there is any price to calculate.

File: tests/api/test_product_minimal_variant_price.py (hunk 15 above)
Comment (maarcingebala) [reply tests/api/test_product_minimal_variant_price.py]: I agree with @korycins here - changes that we add should respect conventions used in the codebase in general. From the name `test_collection_remove_products_updates_minimal_variant_price` you'll also know which operation we're testing, and if that's not enough, look at the query in the test.

**Comment by maarcingebala** [comment]
@koradon I think it is related, but the problem is more general. Consider this case:
1. I have a collection with a bunch of products. The collection is not published.
2. I'm creating a sale and associating it with the above collection.
Now - is the discount active, if the collection is not published yet? Currently, it seems that it is which I consider a bug. But in general - both discounts and collections can be activated/published at some point in the future, and minimal variant price should become valid from that point of time (or should be recalculated at that point).

Maybe the simplest solution is to have a periodic task that recalculates all prices based on what is currently published in the store.

**Comment by koradon** [comment]
@maarcingebala yes, we ware talking about designing periodic task for recalculating prices. That was the main save behavior that will fix any mistakes in our logic.
According to your example:
We should calculate min_price every time when Sale or Collection changes state.

File: saleor/graphql/schema.graphql [hunk 17]
```diff
@@ -2874,7 +2877,7 @@ type Query {
   collection(id: ID!): Collection
   collections(filter: CollectionFilterInput, query: String, before: String, after: String, first: Int, last: Int): CollectionCountableConnection
   product(id: ID!): Product
-  products(filter: ProductFilterInput, attributes: [AttributeScalar], categories: [ID], collections: [ID], priceLte: Float, priceGte: Float, sortBy: ProductOrder, stockAvailability: StockAvailability, query: String, before: String, after: String, first: Int, last: Int): ProductCountableConnection
+  products(filter: ProductFilterInput, attributes: [AttributeScalar], categories: [ID], collections: [ID], priceLte: Float, priceGte: Float, minimalPriceLte: Float, minimalPriceGte: Float, sortBy: ProductOrder, stockAvailability: StockAvailability, query: String, before: String, after: String, first: Int, last: Int): ProductCountableConnection
```
Comment (NyanKiyoshi) [inline saleor/graphql/schema.graphql]: `minimalPriceAmountLte` and `minimalPriceAmountGte`

File: saleor/product/migrations/0104_product_minimal_variant_price.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+        ),
+        migrations.RunPython(
+            populate_product_minimal_variant_price,
+            reverse_code=migrations.RunPython.noop,
+        ),
+        migrations.AlterField(
+            model_name="product",
+            name="minimal_variant_price",
+            field=django_prices.models.MoneyField(
+                currency="USD", decimal_places=2, max_digits=12
```
Comment (NyanKiyoshi) [inline saleor/product/migrations/0104_product_minimal_variant_price.py]: You need to take the currency from `settings.DEFAULT_CURRENCY

File: saleor/product/migrations/0104_product_minimal_variant_price.py [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [("product", "0103_schema_data_enterprise_grade_attributes")]
+
+    operations = [
+        migrations.AddField(
+            model_name="product",
+            name="minimal_variant_price",
+            field=django_prices.models.MoneyField(
+                currency="USD", decimal_places=2, max_digits=12, null=True
```
Comment (NyanKiyoshi) [inline saleor/product/migrations/0104_product_minimal_variant_price.py]: Same, `settings.DEFAULT_CURRENCY`

File: tests/api/test_product_minimal_variant_price.py (hunk 15 above)
Comment (derenio) [reply tests/api/test_product_minimal_variant_price.py]: I've renamed the test to respect the conventions.

File: saleor/product/migrations/0104_product_minimal_variant_price.py [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+import django_prices.models
+from django.db import migrations
+
+
+def populate_product_minimal_variant_price(apps, schema_editor):
+    Product = apps.get_model("product", "Product")
+    for product in Product.objects.iterator():
+        # Set the "minimal variant price" to the default product's price
+        product.minimal_variant_price = product.price
```
[the remark this answers is not in the document]
Comment (derenio) [reply saleor/product/migrations/0104_product_minimal_variant_price.py]: I've added comment in the migration that explains why we don't calculate the correct minimal prices in it (we would have to copy the code from the methods) and suggesting to run the management command `update_all_products_minimal_variant_prices`.

File: saleor/graphql/product/schema.py (hunk 16 above)
Comment (derenio) [reply saleor/graphql/product/schema.py]: I've removed the `minimal_price_(lte|gte)` input fields and also added `DEPRECATION` warning in the `price_(lte|gte)` fields.

File: saleor/graphql/schema.graphql (hunk 17 above)
Comment (derenio) [reply saleor/graphql/schema.graphql]: We are removing these fields.

File: saleor/product/migrations/0104_product_minimal_variant_price.py (hunk 18 above)
Comment (derenio) [reply saleor/product/migrations/0104_product_minimal_variant_price.py]: Thanks, good catch!
These were generated by django and I wouldn't notice it myself.

**Comment by derenio** [comment]
@maarcingebala I've rebased (twice x|) onto the latest master to resolve conflicts on the `Pipfile.lock` and the migration files.

File: saleor/product/models.py [hunk 21]
```diff
@@ -207,6 +236,14 @@ def __repr__(self):
     def __str__(self):
         return self.name
 
+    def save(
+        self, force_insert=False, force_update=False, using=None, update_fields=None
```
Comment (NyanKiyoshi) [inline saleor/product/models.py]: Why not `*args, **kwargs`?

Comment (derenio) [reply saleor/product/models.py]: Because this is the real signature and when you are using/overriding this method you know exactly what arguments it requires.
